### PR TITLE
Dedicated thread pools for sim2h connection handling and message processing

### PR DESCRIPTION
**Pull request by struktured**
## PR summary

 - Makes all the shared fields of `Sim2h` struct into concurrent, thread safe structures (either a `CHashMap` or a `RwLock` strategy with `Arc<_>` guards.
 - Makes a separate thread pool to processing incoming connections (thus *not* in the processing loop any longer). 

## testing/benchmarking notes

( if any manual testing or benchmarking was/should be done, add notes and/or screenshots here )

## followups

( any new tickets/concerns that were discovered or created during this work but aren't in scope for review here )

## changelog

- [ ] if this is a code change that effects some consumer (e.g. zome developers) of holochain core,  then it has been added to [our between-release changelog](https://github.com/holochain/holochain-rust/blob/develop/CHANGELOG-UNRELEASED.md) with the format 

```markdown
- summary of change [PR#1234](https://github.com/holochain/holochain-rust/pull/1234)
```

## documentation

- [ ] this code has been documented according to our [docs checklist](https://hackmd.io/@freesig/Hk9AmKJNS)

### CODE REVIEW

File: crates/sim2h/src/lib.rs [hunk 1]
```diff
@@ -482,15 +601,26 @@ impl Sim2h {
...
+    fn process_internal(&mut self) -> Sim2hResult<()> {
+        trace!("process");
         self.num_ticks += 1;
         if self.num_ticks % 60000 == 0 {
             debug!(".");
             self.num_ticks = 0;
         }
 
-        self.priv_check_incoming_connections();
+        // Now down in separate thread!
```
Comment (struktured) [inline crates/sim2h/src/lib.rs]:
```suggestion
        // Now done in a separate thread!
```

File: crates/sim2h_server/src/main.rs [hunk 2]
```diff
@@ -54,16 +55,18 @@ fn main() {
...
 
-    loop {
-        let result = sim2h.process();
-        if let Err(e) = result {
-            if e.to_string().contains("Bind error:") {
-                println!("{:?}", e);
-                exit(1)
-            } else {
-                error!("{}", e.to_string())
+    for _i in 0..num_cpus::get() {
```
Comment (neonphog) [inline crates/sim2h_server/src/main.rs]: ... this doesn't start any threads or anything, and the infinite loop inside cannot exit. `_i` will never advance beyond `0`.

File: crates/sim2h/src/lib.rs [hunk 3]
```diff
@@ -103,27 +108,34 @@ pub enum DhtAlgorithm {
...
+    connection_states: ConnectionStates,
+    spaces: Spaces,
+    pool: Arc<Pool>,
+    wss_recv: crossbeam_channel::Receiver<TcpWss>,
+    msg_send: crossbeam_channel::Sender<(Url2, FrameResult)>,
+    msg_recv: crossbeam_channel::Receiver<(Url2, FrameResult)>,
+    open_connections: OpenConnections,
     num_ticks: u64,
-    /// when should we recalculated the rrdht_arc_radius
+    /// when should we eecalculated the rrdht_arc_radius
```
Comment (neonphog) [inline crates/sim2h/src/lib.rs]:
```suggestion
    /// when we should recalculate the rrdht_arc_radius
```

File: crates/sim2h/src/lib.rs [hunk 4]
```diff
@@ -179,17 +193,29 @@ impl Sim2h {
...
-                .insert(url, (job.clone(), outgoing_send));
-            self.pool.push_job(Box::new(job));
+        for _i in 0..NUM_CONNECTION_THREADS {
+            let connection_states = self.connection_states.clone();
+            let open_connections = self.open_connections.clone();
+            let msg_send = self.msg_send.clone();
+            let pool = self.pool.clone();
+            let wss_recv = self.wss_recv.clone();
+
+            std::thread::spawn(move || loop {
```
Comment (neonphog) [inline crates/sim2h/src/lib.rs]: I guess sim2h is still using that code that panics if a thread panics... but we still might want to capture the join handle off this spawn / clean it up if / when the sim2h instance gets dropped?

File: crates/sim2h/src/lib.rs [hunk 5]
```diff
@@ -998,16 +1181,26 @@ impl Sim2h {
 
         let payload: Opaque = msg.clone().into();
 
-        match self.open_connections.get_mut(&uri) {
-            None => {
-                error!("FAILED TO SEND, NO ROUTE: {}", uri);
-                return;
-            }
-            Some((_con, outgoing_send)) => {
+        let get_result = open_connections.get_mut(&uri);
```
Comment (neonphog) [inline crates/sim2h/src/lib.rs]: I don't think this is going to do what you want... [get_mut](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.get_mut) returns a None if somebody else currently has a lock, in which case it'll print out the NO ROUTE error message below and exit, which isn't right.

Comment (neonphog) [reply crates/sim2h/src/lib.rs]: n/m it calling the get_mut on the map, not the arc : )

File: crates/sim2h_server/src/main.rs (hunk 2 above)
Comment (struktured) [reply crates/sim2h_server/src/main.rs]: Woops, I blame lack of sleep? Fixed.

File: crates/sim2h/src/lib.rs (hunk 4 above)
Comment (struktured) [reply crates/sim2h/src/lib.rs]: Yes, definitely. Will do in a follow up commit if this code ends up useful in the first place.

**Comment by struktured** [comment]
Important results from this branch, if any, have been incorporated into develop already.

**Pull request closed without merging**